spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/security/oauth2/client/ClientPropertyDefaultsEnvironmentPostProcessor.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2012-2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.autoconfigure.security.oauth2.client;
+
+import java.io.IOException;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.env.EnvironmentPostProcessor;
+import org.springframework.core.Ordered;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.core.io.DefaultResourceLoader;
+import org.springframework.core.io.ResourceLoader;
+import org.springframework.core.io.support.ResourcePropertySource;
+
+/**
+ * An {@link EnvironmentPostProcessor} that registers a <code>PropertySource</code> against the
+ * <code>Environment</code>, which contains a default set of client properties.
+ *
+ * @author Joe Grandja
+ * @since 2.0.0
+ */
+public class ClientPropertyDefaultsEnvironmentPostProcessor implements EnvironmentPostProcessor, Ordered {
```

Comment:
I think this should be replaced with one or more `@ConfigurationProperties` classes with fields initialised with the desired default values.

Reply:
I did look at potentially using `@ConfigurationProperties` initially but it did not suit my requirements because it supports a **static** base property path whereas the base property path for client properties is **dynamic**.

Given the following format for client properties:

`[client-property-prefix].[client-key].[property-name]=[property-value]`

for example, `security.oauth2.client.google.client-id=google-client-id`

the `[client-key]` (google) is **dynamic** and may be any value assigned by the user.

Here's a use case. The user wants to configure the following 2 google clients.

`security.oauth2.client.google-login.client-id=google-login-client-id`
and
`security.oauth2.client.google-calendar.client-id=google-calendar-client-id`

The `google-login` client will be used for `oauth2Login()` whereas the `google-calendar` client will be used for accessing the _Calendar API_.

The user is **not fixed** to using `google` for the `client-key`. They may choose whatever name for `client-key` that makes sense for their configuration.

*NOTE:* The `client-key` is only used as a logical separator for client configurations. Nothing more than that. It is *not used* when auto-configuring `ClientRegistration` instances. Another thing to note as I indirectly mentioned is that the user is able to configure more than one client as they wish - each configuration separated by the `client-key`. This is a new capability in Spring Security 5 compared to `spring-security-oauth` where you are limited to configuring only 1 client.

We're also providing a set of common defaults for well-known providers via `spring-security-oauth2-client-defaults.properties` in order to minimize the configuration for the user. So if the user wants to configure 1 google client only than they can leverage the defaults and would only have to configure the following:

`security.oauth2.client.google.client-id`
`security.oauth2.client.google.client-secret`

However, if they decided to name their `client-key` to `google-login` instead, than they can still leverage the default property values in the `Environment` but their configuration would need to look like this:

`security.oauth2.client.google-login.client-id=google-client-id`
`security.oauth2.client.google-login.client-secret=google-client-secret`
`security.oauth2.client.google-login.authorization-uri=${security.oauth2.client.google.authorization-uri}`
`security.oauth2.client.google-login.token-uri=${security.oauth2.client.google.token-uri}`
...

...leveraging property placeholders.

Hopefully that explains things. And if you have another suggestion for implementation than please let me know.

Reply:
Thanks for the explanation.

You could largely do this with `@ConfigurationProperties` with a `Map<String, Foo>` property where `Foo` would be a pojo with `client-id`, `client-secret` etc properties on it. This is preferable to using the environment directly as it's strongly typed and will provide some auto-completion and documentation in a developer's IDE. I say some, as you won't get any metadata (and therefore IDE support) for anything nested in the map. That situation will improve when we implement https://github.com/spring-projects/spring-boot/issues/9945 but it's out of scope for 2.0.

/cc @snicoll

Reply:
@wilkinsona I'm not sure I follow you? Let's go by example. Given that `@ConfigurationProperties` is allowed on `Type` and `Method` only, these are my 2 options:

**Option 1**
```
	@ConfigurationProperties(value = "security.oauth2.client.google")
	public ClientRegistrationProperties googleClientConfig() {
		return new ClientRegistrationProperties();
	}
```

**Option 2**
```
@ConfigurationProperties(value = "security.oauth2.client.google")
public class GoogleClientConfig extends ClientRegistrationProperties {
}
```

Option 2 is redundant as I'm simply extending the existing `ClientRegistrationProperties` class and provide no extra value here. So this option is not really an option.

Option 1 is an extra configuration step that can be avoided. So for each client I configure in my application.yml or defaults.yml I need to also use `@ConfigurationProperties`. Why? It's really an extra step that IMO is not needed.

Not to mention, as part of the introspection of client properties in the `Environment`, I'm checking for the required `client-id` property and if it's not provided than I don't configure the client. How can I do this validation by-pass with `@ConfigurationProperties`?

Moreover, we do not want to expose `ClientRegistration` instance as beans in the context. The only bean that should be in the context is a `ClientRegistrationRepository` bean that is composed of one or more `ClientRegistration`'s

Reply:
That's not your only two options for what you need to bind. Here is a simple example:

```java
@ConfigurationProperties(value = "security.oauth2")
public class OAuth2Properties {

    private final Map<String,ClientRegistrationProperties> client = new HashMap<>();

    public Map<String,ClientRegistrationProperties> getClient() { return this.client; }

}
```

With this setup, you can bind any client via the map, something like `security.oauth2.client.google.xyz` where `xyz` is a property of `ClientRegistrationProperties`.

Having said that, I'd be in favour of restructing that a bit, something like `security.oauth2.client.providers` or something. The code above is a bit weird, this is much better IMO:

```java
@ConfigurationProperties(value = "security.oauth2.client")
public class OAuth2ClientProperties {

    private final Map<String,ClientRegistrationProperties> providers = new HashMap<>();

    public Map<String,ClientRegistrationProperties> getProviders() { return this.providers; }

}
```

Reply:
@snicoll I agree I do like `OAuth2ClientProperties`. However, I still feel that I need `ClientPropertyDefaultsEnvironmentPostProcessor` because I don't want to load `spring-security-oauth2-client-defaults.properties` into the `Environment` unless there is at least one valid client configured by the user - valid being `client-id` is provided. I can certainly leverage `OAuth2ClientProperties` in `ClientPropertyDefaultsEnvironmentPostProcessor` and get rid of `ClientPropertiesUtil` altogether.

What do you think?

Reply:
Sorry you lost me. I was merely replying to your comment and none of the issues you've raised are applicable to our proposal as far as I can see. 

It feels backwards to do what you are doing in `EnvironmentPostProcessor`. If you need to conditionally load and apply something, you can do that in the auto-configuration I admit I don't understand how those defaults are supposed to be applied: if the idea is that the user has to use some token for it, i.e. `${spring.foo.default.something}` instead of the value, I am not sure I am fan of that idea.

Reply:
@wilkinsona @snicoll Thanks for the comments. I pretty sure I know what I need to do now. I'll push the update shortly which will remove the `EnvironmentPostProcessor`.

Reply:
@wilkinsona @snicoll I can't seem to get the binding to work for `Map<String,ClientRegistrationProperties>`. I wrote a test to prove this out. Can you take a look at this [commit](https://github.com/spring-projects/spring-boot/pull/10235/commits/b2326a219b789c7369ef8fd4430c97f59b0556da) and let me know if you see something out of place?

Reply:
No worries. I got it figured out.

Reply:
@snicoll @wilkinsona I tried your suggestion for binding the clients as follows:

```@ConfigurationProperties(prefix = "security.oauth2.client")
public class OAuth2ClientsProperties {

	private Map<String, ClientRegistrationProperties> clients = new HashMap<>();

	public Map<String, ClientRegistrationProperties> getClients() {
		return clients;
	}

	public void setClients(Map<String, ClientRegistrationProperties> clients) {
		this.clients = clients;
	}
}
```

But this did not work since the base property path used for binding would be `security.oauth2.client.clients` based on this structure but the clients are configured on this base path `security.oauth2.client.<client-key>`, for example, `security.oauth2.client.google`, `security.oauth2.client.github`, etc.

To get it to work, the structure is this:

```
@ConfigurationProperties(prefix = OAuth2ClientsProperties.CLIENT_PROPERTY_PREFIX)
public class OAuth2ClientsProperties extends HashMap<String, ClientRegistrationProperties> {

	public static final String CLIENT_PROPERTY_PREFIX = "security.oauth2.client";

}
```

I will admit that I don't like the fact that I have to extend `HashMap` even though the client properties is a `Map`. As you are reviewing things, maybe you can suggest an alternative option?

Reply:
@jgrandja I gave you a code sample for that, see `OAuth2Properties` (there is no need for a setter if the map is initialized).

As far as I know we decide what the path will be. If we decide that the path is `security.oauth2.client.providers` so bet it. In that case the second example I gave would work.

Sorry, but I don't understand what the problem is.

Reply:
@snicoll Sorry for the confusion but honestly I'm confused as well. In your initial [comment](https://github.com/spring-projects/spring-boot/pull/10235#discussion_r138118003), you did suggest `OAuth2Properties` but shortly after that you said...
> Having said that, I'd be in favour of restructing that a bit, something like security.oauth2.client.providers or something. The code above is a bit weird...

In your last comment, you said
> I think we'd like this to move to spring.security.oauth2.* anyway.

Can you please confirm if you want `spring` at the start of the prefix? This is new and threw me for a loop.

Again, I apologize for the confusion but I want to make sure I get things right so we're not going back and forth here. I don't want to waste your time or mine.

I just pushed an update that includes this:

```
@ConfigurationProperties(prefix = "security.oauth2")
public class OAuth2Properties {

	private Map<String, ClientRegistrationProperties> clients = new LinkedHashMap<>();

	public Map<String, ClientRegistrationProperties> getClients() {
		return this.clients;
	}

	public void setClients(Map<String, ClientRegistrationProperties> clients) {
		this.clients = clients;
	}
}
```

If this is not what you expect than please let me where to correct and I'll get it done quick.
Thanks.

Reply:
I think the back and forth is because the only thing I am doing here is answering you on the technicality. I gave you options that work and answers the technical concern that you've raised.

> Can you please confirm if you want spring at the start of the prefix? This is new and threw me for a loop.

I don't know and it doesn't matter. If/Once the PR is in a good shape to be merged, we can rename the prefix ourselves. It's really not a big deal.

Reply:
@snicoll I'm really hoping we can get this into M4 this Thursday. Please let me know what I need to do on my end to make this happen. Thanks.